Description: Change server name to cloudflare-nginx
Author: hda_launchpad (hda_launchpad) <admin@hda.me>
--- nginx-1.11.11-1.orig/src/http/ngx_http_header_filter_module.c
+++ nginx-1.11.11-1/src/http/ngx_http_header_filter_module.c
@@ -46,9 +46,9 @@ ngx_module_t  ngx_http_header_filter_mod
 };
 
 
-static u_char ngx_http_server_string[] = "Server: nginx" CRLF;
-static u_char ngx_http_server_full_string[] = "Server: " NGINX_VER CRLF;
-static u_char ngx_http_server_build_string[] = "Server: " NGINX_VER_BUILD CRLF;
+static u_char ngx_http_server_string[] = "Server: cloudflare-nginx" CRLF;
+static u_char ngx_http_server_full_string[] = "Server: cloudflare-nginx" CRLF;
+static u_char ngx_http_server_build_string[] = "Server: cloudflare-nginx" CRLF;
 
 
 static ngx_str_t ngx_http_status_lines[] = {
--- nginx-1.13.5-1.orig/src/http/v2/ngx_http_v2_filter_module.c
+++ nginx-1.13.5-1/src/http/v2/ngx_http_v2_filter_module.c
@@ -143,7 +143,7 @@ ngx_http_v2_header_filter(ngx_http_reque
     ngx_http_core_srv_conf_t  *cscf;
     u_char                     addr[NGX_SOCKADDR_STRLEN];
 
-    static const u_char nginx[5] = "\x84\xaa\x63\x55\xe7";
+    static const u_char nginx[13] = "\x8c\x25\x7\xb6\x49\x68\x1d\x85\x5a\xa9\x8d\x57\x9f";
 #if (NGX_HTTP_GZIP)
     static const u_char accept_encoding[12] =
         "\x8b\x84\x84\x2d\x69\x5b\x05\x44\x3c\x86\xaa\x6f";

