Description: Change server name and tls record size
Author: hda_launchpad (hda_launchpad) <admin@hda.me>
--- nginx-1.10.0-1.orig/src/event/ngx_event_openssl.c
+++ nginx-1.10.0-1/src/event/ngx_event_openssl.c
@@ -740,7 +740,7 @@ ngx_ssl_info_callback(const ngx_ssl_conn
             wbio = SSL_get_wbio((ngx_ssl_conn_t *) ssl_conn);
 
             if (rbio != wbio) {
-                (void) BIO_set_write_buffer_size(wbio, NGX_SSL_BUFSIZE);
+                (void) BIO_set_write_buffer_size(wbio, 16384);
                 c->ssl->handshake_buffer_set = 1;
             }
         }
--- nginx-1.10.0-1.orig/src/event/ngx_event_openssl.h
+++ nginx-1.10.0-1/src/event/ngx_event_openssl.h
@@ -135,7 +135,7 @@ typedef struct {
 #define NGX_SSL_BUFFER   1
 #define NGX_SSL_CLIENT   2
 
-#define NGX_SSL_BUFSIZE  16384
+#define NGX_SSL_BUFSIZE  1360
 
 
 ngx_int_t ngx_ssl_init(ngx_log_t *log);
--- nginx-1.11.11-1.orig/src/http/ngx_http_header_filter_module.c
+++ nginx-1.11.11-1/src/http/ngx_http_header_filter_module.c
@@ -46,9 +46,9 @@ ngx_module_t  ngx_http_header_filter_mod
 };
 
 
-static u_char ngx_http_server_string[] = "Server: nginx" CRLF;
-static u_char ngx_http_server_full_string[] = "Server: " NGINX_VER CRLF;
-static u_char ngx_http_server_build_string[] = "Server: " NGINX_VER_BUILD CRLF;
+static u_char ngx_http_server_string[] = "Server: cloudflare-nginx" CRLF;
+static u_char ngx_http_server_full_string[] = "Server: cloudflare-nginx" CRLF;
+static u_char ngx_http_server_build_string[] = "Server: cloudflare-nginx" CRLF;
 
 
 static ngx_str_t ngx_http_status_lines[] = {

